src/network/core/tcp_game.h

```diff
@@ -502,6 +503,13 @@ class NetworkGameSocketHandler : public NetworkTCPSocketHandler {
 	 */
 	virtual NetworkRecvStatus Receive_CLIENT_MOVE(Packet &p);
 
+	/**
+	 * Request the server to authorize a given client to join our company.
+	 * uint32_t  ID of the client to authorize to join our company.
+	 * @param p The packet that was just received.
+	 */
+	virtual NetworkRecvStatus Receive_CLIENT_COMPANY_AUTHORIZE(Packet &p);
```

Comment:
I wonder if this is the best approach; we could also make it a DoCommand, and that way all clients get notified of the change too.

If we would use a DoCommand, the server will need to do filtering on the command ofc. But this isn't something new, we already do that for join-company etc. Basically, it will become comparable with `CMD_COMPANY_CTRL`.

It would also remove the need for `joinable_companies` to exist. And later additions might be easier, as dealing with the Command system from GUIs etc we have plenty of examples for.

Additionally, if using the DoCommand system, and the client makes a savegame at any moment in time, they too have the right access list in their game. As, if I read it correctly, currently it does not. It does receive the initial allow-list, but it is not modified (client-side) after. And for some reason this feels like asking for a bit of trouble.

But in the tldr, it feels there are currently two systems: `joinable_companies` to talk server<->client, and `network_authorized_keys` server-side, which is also sent to the clients on join. It feels like using DoCommand would collapse them into a single system, making it easier to understand.

And now I am curious about your thoughts on this :)

Reply:
To be brutally honest, I do not think that we should actually be sending the public keys of all clients to all clients as that would make tracking clients by non-servers way too trivial to implement. Within the context of servers there's nothing that can be done about that I fear.

The only change that would be required for that is saving an empty collection of authorized keys when we are sending the network saves compared to saving it just to disk.

Reply:
Ah, okay, so you want to keep the public keys of clients hidden as you are afraid people would monitor who is where, etc.

When I worked out how multiplayer could look like, many moons ago (https://gist.github.com/TrueBrain/05d5d066efedd656ec7daa05f92fa012), it kinda was the point that public keys became this identity you can share with others, to have friends list, and to find each other. But that indeed explicitly makes you "trackable" (in some sense) on what server you are etc.

I can see both arguments to this story .. on one hand you have user-friendly-ness, on the other hand you have do-not-track. But given the state you have this PR in, and that it is more about the second than the first, I have no problem going with the second (for now at least :P ), and see where it leads us.

But yes, it does require finding a solution to ensure the authorized-keys are not send to the clients; as that would instantly defeat that purpose :D (you can track by downloading the map; which wouldn't be much different than sending the public-keys to everyone).

Personally, I don't share the hesitation of not sharing the pubkey with everyone. And yes, in some sense that means you can be tracked where you are playing. But I am really not worried someone can see on what server I am playing etc .. it is a game after all. So my acceptance for that is a lot higher. That said, it is my personal opinion, and I have no problem going with yours for now :)

Reply:
Another thought, but really not for this PR: one could implement two client-side modes: "Privacy" and "Public".

With the first, you generate a new pubkey for every server you connect to (stable for that server, but never reused with another server), based on invite-code or something.

With the second, you have a single pubkey for every server.

That would also tackle the "do not track" situation, giving the control to the player what level of privacy they would like.

But again, if you ask me, not for this PR; after this explanation of yours, I am find with this approach :) Maybe others have an opinion :P

Reply:
The command idea is intriguing, but it has some drawbacks to think about:
* just passing authorized keys as string; that works fine, but someone can just keep sending different strings and essentially cause a denial of service attack on servers by either memory exhaustion or blowing up the saves to a point where clients cannot join fast enough.
* using `ClientID`, and resolving the authorized key from there. Now your client is a bit slow, so server executes the command successfully and then sends that the client left. The client will process the "left" message and remove the client, which makes the client not resolve and we have a different state in the server and client. That would mean that moving client moving to commands causes a desync.
* so you would need to send the authorized key and client, add special code for validating those to the server, and then reject or accept that. I'm not sure that's the way to go.

Of course the first case can also be triggered with the other possibilities, but at most MAX_COMPANY times the authorized key for each of the joined clients. As such, it would grow significantly slower as a lot of clients need to join.

Not that I see a really clean and simple way to keep the authorized keys from the save for non locally stored server games.

Reply:
The first problem could be prevented by just adding a max to the length of the authorize list. That is not that weird, I would say. Limit of 50 or so.

I am not sure I fully understand your second and third point. I think you are talking about my suggestion to just share pubkeys with clients/servers? As if so, I think it would make most sense to just send pubkeys, and not ClientIDs. As that way, you can also have a client authorize a pubkey of a player not in-game currently. I cannot see any scenario where you also want to send a ClientID there, or the server need to do any validation if they match. I would just keep it simple.

There are a bunch of settings you could add here, ofc. For example, a setting that does validate if it knows the pubkey. But this too is irrelevant to the ClientID. 

If you do want to use the ClientID for this DoCommand, I think that only makes sense in a scenario where you want to keep the pubkey server-side only. But in that case, using DoCommand is a bit tricky, as how do you tell a client that is joining which company it can join?

Either way, I do not really understand what you mean with your second drawback.

- Client A sends DoCommand with ClientID
- Server processes DoCommand, adds pubkey of ClientID to allow-list
- Server broadcasts DoCommand to all other clients
- Server notices client A is too slow, and kicks it
- Client A is now irrelevant, it is disconnected
- Client B...Z receive the DoCommand to add the pubkey to the allow-list
- Client B..Z receive disconnect

I guess you talk about the situation where the last and second-to-last are switched around? Another reason to not use ClientID when you share pubkeys between client/server, and irrelevant if you keep the pubkey server-side-only as the only one who actually cared was the client that is disconnected :)

Reply:
I have changed it to a command in an effort to keep the allow lists of the companies the same between clients and server